#### Theory

Household utility is determined by health $h$ as a function of food production $f_i\ \forall\ i \in I$ from available means and the composite good $\phi$. Each food is a comprised of protein $y_P$, fat $y_F$, and carbohydrate $y_C$ macronutrients. Formally, utility is expressed as $U = U[h(f_1(y_P, y_F, y_C)), h(f_j(y_P, y_F, y_C)),..., h(f_I(y_P, y_F, y_C)), \phi]$. It is assumed that utility and production are differentiable, concave, and that food production in utility is weakly separable. Household market demand is provided by $y_m$ for $m = \{P, F, C\}$, income $I$, prices $p_m$, and a basket of household characteristics $\theta$. Formally, household demand is expressed as $y_m = y_m(I, p_m, \theta)$.

Muth (1966) considers household demand of commodity $y_i\ \forall\ i \in n$ for household production $x_g = x_g(y_{gi},...,y_{gn})\ \forall\ g \in G$, with household utility defined as $U = U(x_1,...,x_G)$. Evaluating compensated substitution effects, the demand for $y_i$ is defined as $y_i = y_i(I, p_j)$ for the price $p$ of commodity $j$. The compensated cross price effect can be characterized as 

\begin{equation*}
\frac{\partial y_i^c}{\partial p_j} = K^{gh}\frac{\partial y_i}{I}\frac{\partial y_j}{I},
\end{equation*}

for $i \in g$ and $j \neq i \in h$. $K^{gh}$ for $g \neq h$ is defined as a function of the quantities of initial commodity positions in the household's bundle. Converting equation (1) to represent elasticities is useful for policy impact evaluations. Manipulation of the expression above provides

\begin{align*}
\frac{\partial y_i^c}{\partial p_j}\left(\frac{p_j}{y_i}\right) &= \left[K^{gh}\frac{\partial y_i}{\partial I}\left(\frac{I}{I}\right)\left(\frac{y_i}{y_i}\right)\frac{\partial y_j}{\partial I}\left(\frac{I}{I}\right)\left(\frac{y_j}{y_j}\right)\right]\left(\frac{p_j}{y_i}\right) \\[7pt]
&\Rightarrow \left[\frac{K^{gh}}{I}\frac{\partial y_i}{\partial I}\frac{I}{y_i}y_i\frac{\partial y_j}{\partial I}\frac{I}{y_j}\frac{y_j}{I}\right]\left(\frac{p_j}{y_i}\right) \\[7pt]
&\Rightarrow \left[\frac{K^{gh}}{I}E_iE_j\frac{y_iy_j}{I}\right]\left(\frac{p_j}{y_i}\right) \\[7pt]
&\Rightarrow \frac{K^{gh}}{I}E_iE_j\frac{p_jy_j}{I} \\[7pt]
&\Rightarrow \frac{K^{gh}}{I}E_iE_js_j = \frac{\partial y_i^c}{\partial p_j}\frac{p_j}{y_i} = H_{ij},
\end{align*}

where $E_i$ and $E_j$ represent price elasticities, $s_j$ represents expenditure on commodity $j$ relative to income $I$, and $H_{ij}$ represents the elasticity of demand for commodity $i$ with respect to changes in the price $p$ of commodity $j$.

Total differentiation of log linearized demand $\text{ln}y_i = \text{ln}y_i(I, p_j)$ provides a functional demand form in terms of income and cross-price elasticities. 

\begin{align*}
\frac{1}{y_i}\text{d}y_i &= \frac{1}{y_i}\frac{\partial y_i}{\partial I}\text{d}I + \frac{1}{y_i}\frac{\partial y_i}{\partial p_j}\text{d}p_j \\[7pt]
&\Rightarrow \frac{1}{y_i}\frac{\partial y_i}{\partial I}\frac{I}{I}\text{d}I + \frac{1}{y_i}\frac{\partial y_i}{\partial p_j}\frac{p_j}{p_j}\text{d}p_j \\[7pt]
&\Rightarrow \frac{\partial y_i}{\partial I}\frac{I}{y_i}\frac{\text{d}I}{I} + \frac{\partial y_i}{\partial p_j}\frac{p_j}{y_i}\frac{\text{d}p_j}{p_j} \\[7pt]
&\Rightarrow E_i\text{d}I^* + \sum_{j \neq i}^JH_{ij}\text{d}p_j^* = \text{d}y_i^*,
\end{align*}

providing a demand function $\text{d}y_i^*$ for commodity $i$. Extending the theoretical demand function framework from Muth (1966) to the representative Kenyan household demand $y_m = y_m(I, p_m, \theta, \varepsilon)$, while taking into account own- as well as cross-price elasticities, also provides a functional demand form in terms of estimable elasticities. It directly follows that

\begin{equation}
\text{d}y_m^* = E_i\text{d}I^* + \eta_m\text{d}p_m^* + \sum_{j \neq m}^MH_{ij}\text{d}p_j^* + \xi_{\theta}\text{d}\theta^*\ \forall\ m = \{P, F, C\} \tag{1} \label{eq1},
\end{equation}

where $\eta_m$ and $\xi_{\theta}$ represent macronutrient own-price and household characteristic demand elasticities. While macronutrient prices do not explicitly exist, shadow prices are derived as the price share of the macronutrient's composition in the food item relative to the other macronutrients, $p_m = \frac{y_m}{f_i}$ for food $i$.

#### Data

PBASS data used for this study consists of 15,326 observations on household food consumption, livestock and crop production, socio-economic indicators, household demographics, and environmental characteristics. Household food consumption is measured by a seven day recall period where the individual provides intake frequencies for various food items the household consumes. Macronutrient conversion is needed to characterize food consumption as macronutrient consumption. Using the strategy set forth in Schmidhuber (2018), macronutrient composition values are found through the United States Department of Agriculture Food Composition Databases. Macronutrient values per 100g of food item consumed are used as inputs into an algorithm that converts household food intake information to macronutrient consumption information. This process is shown below.

In [1]:
### Rural Kenya Nutrient Demand Analysis ###

import numpy as np
import pandas as pd
import glob
import plotly.plotly as py
import plotly.graph_objs as go
from scipy import stats
import scipy.optimize as optim
import warnings
warnings.simplefilter('ignore')

### Data management ###

def files(sub_dir):
    return glob.glob('/home/ajkappes/' + sub_dir + '*.csv')

nutrient_dfs = np.array(files('/Research/Africa/Nutrient_Demand/'))
#print(nutrient_dfs)

df_consump = pd.read_csv(nutrient_dfs[0])
df_nutrient_props = pd.read_csv(nutrient_dfs[1])
df_land_crop = pd.read_csv(nutrient_dfs[2])
df_hh_demographs = pd.read_csv(nutrient_dfs[3])
df_livinc = pd.read_csv(nutrient_dfs[4])
df_hh_asset = pd.read_csv(nutrient_dfs[5])

# Need to match HH_id by month due to repeating/missing id values across years
# then concatenate all months Feb 13 - Jul 16

m_y = pd.DataFrame(df_consump['date'].unique().tolist()).rename(columns={0: 'date'})

d={}
for month_year in m_y['date']:
    d[month_year] = pd.DataFrame()

for key in d:

    d[key] = df_consump[df_consump['date'] == key].drop_duplicates(['HousehldID'], keep='last').merge(

        df_hh_asset[df_hh_asset['date'] == key].drop_duplicates(['HousehldID'], keep='last'),
        how='inner', on='HousehldID').merge(

        df_hh_demographs[df_hh_demographs['date'] == key].drop_duplicates(['HousehldID'], keep='last'),
        how='inner', on='HousehldID').merge(

        df_land_crop[df_land_crop['date'] == key].drop_duplicates(['HousehldID'], keep='last'),
        how='inner', on='HousehldID').merge(

        df_livinc[df_livinc['date'] == key].drop_duplicates(['HousehldID'], keep='last'),
        how='inner', on='HousehldID').dropna()


df_list = [d[m_y.loc[0, 'date']]]
i = 1
while i < len(d):
    df_list.append(d[m_y.loc[i, 'date']])
    i += 1
    if i > len(d):
        break

df = pd.concat(df_list, axis=0).reset_index().drop(columns='index') # Ending with 15326 observations

# Write algorithm to convert food consumption values to macronutrient proportions
# using USDA Food Composition Databases https://ndb.nal.usda.gov/ndb/

consump_l = ['TotalCowMilkConsumed', 'TotalGoatMilkConsumed', 'TotalEggsConsumed', 'TotalBeefConsumed',
             'TotalGoatMeatConsumed', 'TotalOtherMeatConsumed', 'TotalFishConsumed', 'TotalMaizeConsumed',
             'TotalCassavaConsumed', 'TotalSorghumConsumed', 'TotalBananaConsumed', 'TotalPulsesConsumed',
             'TotalViungoConsumed', 'TotalGreensConsumed', 'TotalPotatoConsumed', 'TotalOilConsumed']

consumption = df[consump_l]

# Converting '300ML' string in oil consumption to L float value and 2kg in pulses to 2
consumption.loc[consumption[consumption['TotalOilConsumed'] == '300ML'].index, 'TotalOilConsumed'] = 0.3
consumption.loc[consumption[consumption['TotalPulsesConsumed'] == '2kg'].index, 'TotalPulsesConsumed'] = 2
consumption = consumption.astype('float64')


macronutrients = ['protein', 'fat', 'carb']

goat_sheep_milk = pd.DataFrame(df_nutrient_props[df_nutrient_props['food_source'].isin(['milk_sheep', 'milk_goat'])]
                               [macronutrients].mean()).T

pulses = pd.DataFrame(df_nutrient_props[df_nutrient_props['food_source'].isin(['peas', 'navy', 'pinto',
                                                                               'black', 'kidney', 'lentils'])]
                      [macronutrients].mean()).T

viungo = pd.DataFrame(df_nutrient_props[df_nutrient_props['food_source'].isin(['onions', 'tomatoes', 'carrots',
                                                                               'green_peppers'])]
                      [macronutrients].mean()).T

greens = pd.DataFrame(df_nutrient_props[df_nutrient_props['food_source'].isin(['spinach', 'cabbage'])]
                      [macronutrients].mean()).T

nutrient_comps = pd.concat([df_nutrient_props[df_nutrient_props['food_source'] == 'milk_cow'][macronutrients],
                            goat_sheep_milk,
                            df_nutrient_props[df_nutrient_props['food_source'] == 'eggs'][macronutrients],
                            df_nutrient_props[df_nutrient_props['food_source'] == 'beef'][macronutrients],
                            df_nutrient_props[df_nutrient_props['food_source'] == 'goat'][macronutrients],
                            df_nutrient_props[df_nutrient_props['food_source'] == 'other_poultry'][macronutrients],
                            df_nutrient_props[df_nutrient_props['food_source'] == 'fish'][macronutrients],
                            df_nutrient_props[df_nutrient_props['food_source'] == 'maize'][macronutrients],
                            df_nutrient_props[df_nutrient_props['food_source'] == 'cassava'][macronutrients],
                            df_nutrient_props[df_nutrient_props['food_source'] == 'sorghum'][macronutrients],
                            df_nutrient_props[df_nutrient_props['food_source'] == 'banana'][macronutrients],
                            pulses, viungo, greens,
                            df_nutrient_props[df_nutrient_props['food_source'] == 'potato'][macronutrients],
                            df_nutrient_props[df_nutrient_props['food_source'] == 'oil'][macronutrients]],
                           axis=0).reset_index().drop(columns='index')

# Macronutrient proportions are values per 100g of food item consumed
# conversions: 1000g[water] per L
#              1000g per Kg
#
# measured in L: cow and goat/sheep milk
# measured in Kg: all meat and crop items
# eggs measured in # consumed: medium egg = 44g => scale 100g egg nutrient values by 0.44

conversion = np.array([10, 10, 0.44, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10])
consumption = np.array(consumption)
nutrient_comps = np.array(nutrient_comps)

macros = pd.DataFrame({'protein_cons': np.zeros(len(consumption)),
                       'fat_cons': np.zeros(len(consumption)),
                       'carbs_cons': np.zeros(len(consumption))})


for i in range(consumption.shape[0]):
    macros.loc[i, 'protein_cons'] = np.dot(np.multiply(nutrient_comps[:, 0], consumption[i]), conversion)
    macros.loc[i, 'fat_cons'] = np.dot(np.multiply(nutrient_comps[:, 1], consumption[i]), conversion)
    macros.loc[i, 'carbs_cons'] = np.dot(np.multiply(nutrient_comps[:, 2], consumption[i]), conversion)

Descriptive macronutrient mean statistics are computed and shown below in order to evaluate preliminary seasonality effects.

In [2]:
macro_idx = {}
for month_year in m_y['date']:
    macro_idx[month_year] = pd.DataFrame()

for key in macro_idx:
    macro_idx[key] = df[df['date'] == key].index.tolist()

d_macros = {}
for month_year in m_y['date']:
    d_macros[month_year] = pd.DataFrame()

for key in d_macros:
    d_macros[key] = macros.loc[macro_idx[key]].mean()

df_macro_statlist = [d_macros[m_y.loc[0, 'date']]]
i = 1
while i < len(d_macros):
    df_macro_statlist.append(d_macros[m_y.loc[i, 'date']])
    i += 1
    if i > len(d_macros):
        break

df_macrostat = pd.DataFrame(pd.concat(df_macro_statlist, axis=1).T).round(3)
df_macrostat.index = m_y['date']

# macronutrient means plot
x_dates = df_macrostat.index

trace0 = go.Scatter(
    x=x_dates,
    y=df_macrostat['protein_cons'],
    mode='lines',
    name='Protein'
)

trace1 = go.Scatter(
    x=x_dates,
    y=df_macrostat['fat_cons'],
    mode='lines',
    name='Fat'
)

trace2 = go.Scatter(
    x=x_dates,
    y=df_macrostat['carbs_cons'],
    mode='lines',
    name='Carbohydrates'
)

line_data = [trace0, trace1, trace2]
layout = dict(title='Figure 1. Mean Macronutrient Consumption (7 day periods)',
              yaxis=dict(title='Macronutrient Consumption in Grams'))

figure = dict(data=line_data, layout=layout)
py.iplot(figure, filename='Macronutrient_means_plot')

Figure 1 reveals that carbohydrates constitute the main macronutrient consumed. Corresponding to the beginning of Kenya's general harvest season in October, carbohydrate consumption increases for each yearly period. Shortly before general harvest, protein consumption also increases. Fat consumption increases drastically during the 2014 harvest season but does not have the same effect during other harvest seasons. All macronutrient consumption trends show evidence of variability around the general harvest season. These seasonality effects are quantitatively analyzed using the empirical methods outlined in the following section.

#### Empirical Methods

The elasticities in the functional demand form in equation \eqref{eq1} can be estimated using the Almost Ideal Demand System (AIDS) specification in Deaton and Muellbauer (1980). A slight modification of the base AIDS specification

\begin{equation*}
w_i = \alpha_i + \sum_j\gamma_{ij}\text{ln}p_j + \beta_i\text{ln}\frac{x}{P}
\end{equation*}

yields

\begin{equation}
y_{im} = \alpha_i + \beta_iI + \boldsymbol{\upsilon\theta} + \sum_m\gamma_m\text{ln}p_m + \varphi_i\text{ln}\frac{x}{P} \tag{2} \label{eq2}
\end{equation}

where $x$ represents expenditure on macronutrient $y_m$ and $P$ represents a price index defined as $\text{ln}P = \alpha_0 + \sum_m\alpha_m\text{ln}p_m + \frac{1}{2}\sum_m\sum_{j \neq m}\gamma_{mj}\text{ln}p_m\text{ln}p_j$. Manipulation of the estimable parameters in equation \eqref{eq2}, the macronutrient demand AIDS specification, provides elasticity values for the theoretical demand equation \eqref{eq1}.

Another empirical approach follows LaFrance (1990) in modeling an incomplete demand system, where the realization of having incomplete information is not assumed away. This modeling approach allows for measurement error when analyzing observational data by specifying an error term that represents such error. Demand in this setting is specified as being linear and quadratic in prices

\begin{equation*}
y_{im} = \alpha_i(\mathbf{q}) + \sum_m\gamma_mp_m + \frac{1}{2}\sum_m\sum_j\delta_{mj}p_mp_j + \beta_iI + \boldsymbol{\upsilon\theta},
\end{equation*}

where $\mathbf{q}$ represents the price vector of unobserved $\mathbf{z}$ commodity consumption. In estimable form the household's indirect utility for observable consumption $y_m$ is provided by 

\begin{equation*}
v(\mathbf{p}, \mathbf{q}, I) = I - \boldsymbol{\alpha}'\mathbf{p} - \frac{1}{2}\mathbf{p}'\boldsymbol{\beta}\mathbf{p} - \boldsymbol{\omega}(\mathbf{q})\text{exp}\{-\boldsymbol{\gamma}'\mathbf{p}\}.
\end{equation*}

Using Roy's identity (1947) to derive Marshillian demand $y_m^\text{x}$ gives the estimable demand expression

\begin{equation*}
-\frac{\partial v(\mathbf{p}, \mathbf{q}, I)/ \partial \mathbf{p}}{\partial v(\mathbf{p}, \mathbf{q}, I)/ \partial \mathbf{I}} = y_m^\text{x} = \mathbf{\alpha} + \boldsymbol{\beta}\mathbf{p} + \boldsymbol{\gamma}\left[\boldsymbol{\omega}(\mathbf{q})\text{exp}\{-\boldsymbol{\gamma}'\mathbf{p}\}\right].
\end{equation*}

However, it must be noted that the expression for $y_m^\text{x}$ incorporates no wealth effects $I$. With $\mathbf{q}$ being unobservable from unobserved consumption $\mathbf{z}$, the incomplete demand system specification eludes to the stochastic error term being defined as $\boldsymbol{\varepsilon} = \boldsymbol{\gamma}\left[\boldsymbol{\omega}(\mathbf{q})\text{exp}\{-\boldsymbol{\gamma}'\mathbf{p}\}\right]$. Relevant wealth effects are also incorporated into the error term creating opportunity for omitted variable bias. A linear wealth term can easily be incorporated into the specification for $y_m^{\text{x}}$ but the unobserved effects of $\mathbf{q}$ must be instrumented in order to account for variation in $y_m^{\text{x}}$ explained by the range of household consumption decisions. Evaluating the expression for indirect utility we can say that consumption of $y_m$ takes place only while indirect utility is nonnegative.

\begin{equation*}
v(\mathbf{p}, \mathbf{q}, I) = I - \boldsymbol{\alpha}'\mathbf{p} - \frac{1}{2}\mathbf{p}'\boldsymbol{\beta}\mathbf{p} - \boldsymbol{\omega}(\mathbf{q})\text{exp}\{-\boldsymbol{\gamma}'\mathbf{p}\} \geq 0.
\end{equation*}

At the corner where indirect utility from consumption $y_m$ is zero, the instrument for consumption of unobserved quantities can be provided as

\begin{equation*}
\boldsymbol{\omega}(\mathbf{q})\text{exp}\{-\boldsymbol{\gamma}'\mathbf{p}\} = I - \boldsymbol{\alpha}'\mathbf{p} - \frac{1}{2}\mathbf{p}'\boldsymbol{\beta}\mathbf{p}.
\end{equation*}

This instrument is a linear combination of observed variables, of which includes wealth effects. Using the instrument in the expression for $y_m^{\text{x}}$ provides

\begin{equation}
y_m^{\text{x}} = \mathbf{\alpha} + \boldsymbol{\beta}\mathbf{p} + \boldsymbol{\gamma}\left[I - \boldsymbol{\alpha}'\mathbf{p} - \frac{1}{2}\mathbf{p}'\boldsymbol{\beta}\mathbf{p}\right] = \mathbf{\alpha} + \boldsymbol{\beta}\mathbf{p} + \boldsymbol{\gamma}\hat{\mathbf{\pi}} \tag{3} \label{eq3}
\end{equation}

While wealth effects are not directly observed in the intrument's design, a nonlinear estimation procedure can be employed to directly incorporate wealth effects for the middle term in equation \eqref{eq3} for $y_m^{\text{x}}$. Elasticity estimates can be readily derived from equation \eqref{eq3} and used for the theoretical demand equation \eqref{eq1}.

Empirical methods are evaluated below. It is important to note that household demographic variables are not yet included in the following specified equations. These estimation procedures follow the equations set forth in Deaton and Muellbauer (1988), Moschini (1995), and LaFrance (1990), in order to make sure the nonlinear almost ideal demand system, the linear almost ideal demand system with "corrected" Stone index, and the nonlinear incomplete demand system procedures can be coded and evaluated successfully.

Empirical methods begin with macronutrient shadow price construction and nonlinear almost ideal demand system estimation.

In [19]:
# shadow price construction
# each macronutrient's proportion of total food expense

food_exp_list = [var for var in df_consump.columns if 'Cost' in var][:-4] # removing non-food exps
df['total_fd_exp'] = df_consump[food_exp_list].sum(axis=1) # total cost of food across all foods consumed

macro_props = ['protein_prop', 'fat_prop', 'carb_prop']
for i in range(len(macro_props)):
    macros[macro_props[i]] = macros[macros.columns[i]] / macros[['protein_cons', 'fat_cons', 'carbs_cons']
                                                               ].sum(axis=1)

macro_cost = ['protein_cost', 'fat_cost', 'carb_cost']
for i in range(len(macro_cost)):
    macros[macro_cost[i]] = macros[macro_props[i]] * df['total_fd_exp']

macro_price = ['protein_p', 'fat_p', 'carb_p']
for i in range(len(macro_price)):
    macros[macro_price[i]] = df['total_fd_exp'] / macros[macros.columns[i]]

# almost ideal demand system, deaton and muellbauer (1980)
# the following provides nonlinear estimation by assumed normal MLE or GMM with e ~ N(0, Sigma)

macros = macros.replace([np.inf, -np.inf], np.nan).dropna()

p_df = macros[['protein_p', 'fat_p', 'carb_p']]
p_df = np.log(p_df.loc[~(p_df == 0).all(axis=1)])
p_df.columns = ['lnprotein_p', 'lnfat_p', 'lncarb_p']

def d_sum(var):
    return p_df[var] * p_df[['lnprotein_p', 'lnfat_p', 'lncarb_p']].sum(axis=1)

p_df['lnprotein_sum'] = d_sum('lnprotein_p')
p_df['lnfat_sum'] = d_sum('lnfat_p')
p_df['lncarb_sum'] = d_sum('lncarb_p')

X = np.concatenate([np.array(p_df[['lnprotein_p', 'lnfat_p', 'lncarb_p']]),
                    np.array(df.loc[p_df.index.tolist(), 'total_fd_exp']).reshape(len(p_df), 1),
                    np.array(p_df[['lnprotein_sum', 'lnfat_sum', 'lncarb_sum']])],
                   axis=1)

y = np.array(macros.loc[p_df.index.tolist(), ['protein_prop', 'fat_prop', 'carb_prop']])

# function specified by deaton and muellbauer (1980)

def fun_fit_aids(p, X, y, indicate):
    # alpha params 0-3
    # beta param 4
    # gamma param 5-9
    if indicate == 'protein':
        return ((p[1] - p[4] * p[0]) + p[5] * X[:, 0] + p[6] * X[:, 1] + p[7] * X[:, 2] +
                p[4] * (X[:, 3] - (p[2] * X[:, 1] + p[3] * X[:, 2]) - 0.5 * (p[8] * X[:, 5] + p[9] * X[:, 6]))
                ) - y[:, 0]

    elif indicate == 'fat':
        return ((p[1] - p[4] * p[0]) + p[5] * X[:, 0] + p[6] * X[:, 1] + p[7] * X[:, 2] +
                p[4] * (X[:, 3] - (p[2] * X[:, 0] + p[3] * X[:, 2]) - 0.5 * (p[8] * X[:, 4] + p[9] * X[:, 6]))
                ) - y[:, 1]

    else:
        return ((p[1] - p[4] * p[0]) + p[5] * X[:, 0] + p[6] * X[:, 1] + p[7] * X[:, 2] +
                p[4] * (X[:, 3] - (p[2] * X[:, 0] + p[3] * X[:, 1]) - 0.5 * (p[8] * X[:, 4] + p[9] * X[:, 5]))
                ) - y[:, 2]

def jacobian_aids(p, X, y, indicate):
    j = np.empty((X.shape[0], p.size))

    if indicate == 'protein':
        j[:, 2] = -p[4] * X[:, 1]
        j[:, 3] = -p[4] * X[:, 2]
        j[:, 8] = -p[4] * 0.5 * X[:, 5]
        j[:, 9] = -p[4] * 0.5 * X[:, 6]

    elif indicate == 'fat':
        j[:, 2] = -p[4] * X[:, 0]
        j[:, 3] = -p[4] * X[:, 2]
        j[:, 8] = -p[4] * 0.5 * X[:, 4]
        j[:, 9] = -p[4] * 0.5 * X[:, 6]

    else:
        j[:, 2] = -p[4] * X[:, 0]
        j[:, 3] = -p[4] * X[:, 1]
        j[:, 8] = -p[4] * 0.5 * X[:, 4]
        j[:, 9] = -p[4] * 0.5 * X[:, 5]

    j[:, 0] = -p[4]
    j[:, 1] = 1
    j[:, 4] = -p[0] + X[:, 3] - (p[2] * X[:, 1] + p[3] * X[:, 2]) - 0.5 * (p[8] * X[:, 5] + p[9] * X[:, 6])
    j[:, 5] = X[:, 0]
    j[:, 6] = X[:, 1]
    j[:, 7] = X[:, 2]
    return j

p_init = np.repeat(1, 10)
fit_aids_protein = optim.least_squares(fun_fit_aids, p_init, jac=jacobian_aids, args=(X, y, 'protein'))
fit_aids_fat = optim.least_squares(fun_fit_aids, p_init, jac=jacobian_aids, args=(X, y, 'fat'))
fit_aids_carb = optim.least_squares(fun_fit_aids, p_init, jac=jacobian_aids, args=(X, y, 'carb'))

# inference

#n = X.shape[0]
#k = p_init.size
#sighat2 = (np.matrix(fit_aids.fun) * np.matrix(fit_aids.fun).T)[0, 0] / (n - k)
#J = np.matrix(fit_aids.jac)
#param_cov = sighat2 * np.linalg.inv(J.T * J)
#param_se = np.sqrt(np.diag(param_cov))
#param_tvals = np.divide(fit_aids.x, param_se)
#t_crit = stats.t.ppf(.975, n - k)
#param_pvals = 2 * (1 - stats.t.cdf(abs(param_tvals), n - k))

print('Nonlinear AIDS dependent: protein budget share; price effects:', np.round(fit_aids_protein.x[5:8], 4))
print()
print('Nonlinear AIDS dependent: fat budget share; price effects:', np.round(fit_aids_fat.x[5:8], 4))
print()
print('Nonlinear AIDS dependent: carb budget share; price effects:', np.round(fit_aids_carb.x[5:8], 4))

Nonlinear AIDS dependent: protein budget share; price effects: [-0.1481  0.0293  0.1083]

Nonlinear AIDS dependent: fat budget share; price effects: [ 0.0332 -0.118   0.1092]

Nonlinear AIDS dependent: carb budget share; price effects: [ 0.1046  0.0953 -0.2072]


The first price effect element refers to protein, second refers to fat, and third refers carbohydrate. Values are interpreted as marginal changes in macronutrient budget share. The own-price effect for protein is estimated as -0.148, fat is -0.118, and carbohydrate is -0.207. Cross-price effects are shown as positive. 

OLS almost ideal demand system estimation using the corrected Stone index are shown below.

In [20]:
# corrected stone index for OLS almost ideal demand system, giancarlo moschini (1995)

price_bidx = df.loc[p_df.index.tolist(), 'date'][df['date'] == 'Feb-13'].index.tolist()
price_base = np.empty((len(macro_price)))

for i in range(len(macro_price)):
    price_base[i] = macros.loc[price_bidx, macro_price[i]].mean()

cStone_pidx = ['Sprotein_idx', 'Sfat_idx', 'Scarb_idx']
for i in range(len(cStone_pidx)):
    p_df[cStone_pidx[i]] = macros.loc[p_df.index.tolist(), macro_props[i]] * (p_df[p_df.columns[i]] -
                                                                           np.log(price_base[i]))

X_cStoneidx = np.concatenate([np.ones(len(p_df)).reshape(len(p_df), 1),
                              np.array(p_df[['lnprotein_p', 'lnfat_p', 'lncarb_p']]),
                              np.array(df.loc[p_df.index.tolist(), 'total_fd_exp']).reshape(len(p_df), 1),
                              np.array(p_df[['Sprotein_idx', 'Sfat_idx', 'Scarb_idx']])],
                             axis=1)

def ols_est(X, y, indicate):
    y = np.asmatrix(y).T
    X = np.asmatrix(X)

    if indicate == 'params':
        return np.linalg.inv(X.T * X) * X.T * y

    elif indicate == 'resids':
        return y - X * np.linalg.inv(X.T * X) * X.T * y

    else:
        print('Indicate "parameters" or "residuals"')

ols_params_protein = ols_est(X_cStoneidx, y[:, 0], 'params')
ols_params_fat = ols_est(X_cStoneidx, y[:, 1], 'params')
ols_params_carb = ols_est(X_cStoneidx, y[:, 2], 'params')

# ols inference

#ols_resids = ols_est(X_cStoneidx, y[:, 0], 'resids')
#ols_sighat2 = ((ols_resids.T * ols_resids) / (X_cStoneidx.shape[0] - X_cStoneidx.shape[1]))[0, 0]
#ols_paramcov = ols_sighat2 * np.linalg.inv(np.asmatrix(X_cStoneidx).T * np.asmatrix(X_cStoneidx))
#ols_paramse = np.sqrt(np.diag(ols_paramcov))
#ols_tvals = np.divide(np.array(ols_params).reshape(1, X_cStoneidx.shape[1]), ols_paramse)
#ols_pvals = 2 * (1 - stats.t.cdf(abs(ols_tvals), X_cStoneidx.shape[0] - X_cStoneidx.shape[1]))

print('OLS AIDS dependent: protein budget share; price effects:', np.round(ols_params_protein[1:4].reshape(1, 3), 4))
print()
print('OLS AIDS dependent: fat budget share; price effects:', np.round(ols_params_fat[1:4].reshape(1, 3), 4))
print()
print('OLS AIDS dependent: carb budget share; price effects:', np.round(ols_params_carb[1:4].reshape(1, 3), 4))

OLS AIDS dependent: protein budget share; price effects: [[-0.1296  0.0219  0.0987]]

OLS AIDS dependent: fat budget share; price effects: [[ 0.0231 -0.0985  0.1313]]

OLS AIDS dependent: carb budget share; price effects: [[ 0.1066  0.0766 -0.2301]]


Estimated price effect similarities are shown for the nonlinear and linear almost ideal demand systems routines. The price effect vector elements remain the same: first element corresponds to protein, second element corresponds to fat, and the third elemend corresponds to carbohydrate.

Incomplete demand system estimation is shown below.

In [23]:
# nonlinear estimation using LaFrance (1990) incomplete demand specification

# converted str #VALUE! error in crop income to zero
df.loc[df.loc[p_df.index.tolist()][df['cropincome'] == '#VALUE!'].index.tolist(), 'cropincome'] = 0

p_df['total_inc'] = df.loc[p_df.index.tolist(),
                           [var for var in df.columns if 'ncome' in var]].astype('float64').sum(axis=1)

def m_sum(var):
    return macros[var] * macros[macro_price].sum(axis=1)

macros['protein_psum'] = m_sum('protein_p')
macros['fat_psum'] = m_sum('fat_p')
macros['carb_psum'] = m_sum('carb_p')

X_ids = np.concatenate([np.array(macros.loc[p_df.index.tolist(), macro_price]),
                        np.array(p_df['total_inc']).reshape(len(p_df), 1),
                        np.array(macros.loc[p_df.index.tolist(), ['protein_psum', 'fat_psum', 'carb_psum']])
                        ], axis=1)

y_ids = np.array(macros.loc[p_df.index.tolist(), ['protein_cons', 'fat_cons', 'carbs_cons']])

def fun_fit_ids(p, X, y, indicate):
    # alpha param 0-3
    # beta params 4-9
    # gamma 10-12

    if indicate == 'protein':
        dep = y[:, 0]

    elif indicate == 'fat':
        dep = y[:, 1]

    else:
        dep = y[:, 2]

    return (p[0] + p[4] * X[:, 0] + p[5] * X[:, 1] + p[6] * X[:, 2] +
            p[10] * (X[:, 3] - (p[1] * X[:, 0] + p[2] * X[:, 1] + p[3] * X[:, 2]) -
                     0.5 * (p[7] * X[:, 4] + p[8] * X[:, 5] + p[9] * X[:, 6])) +
            p[11] * (X[:, 3] - (p[1] * X[:, 0] + p[2] * X[:, 1] + p[3] * X[:, 2]) -
                     0.5 * (p[7] * X[:, 4] + p[8] * X[:, 5] + p[9] * X[:, 6])) +
            p[12] * (X[:, 3] - (p[1] * X[:, 0] + p[2] * X[:, 1] + p[3] * X[:, 2]) -
                     0.5 * (p[7] * X[:, 4] + p[8] * X[:, 5] + p[9] * X[:, 6]))
            ) - dep

def jacobian_ids(p, X, y, indicate):
    j = np.empty((X_ids.shape[0], p.size))
    j[:, 0] = 1
    j[:, 1] = -p[10] * X[:, 0] - p[11] * X[:, 0] - p[12] * X[:, 0]
    j[:, 2] = -p[10] * X[:, 1] - p[11] * X[:, 1] - p[12] * X[:, 1]
    j[:, 3] = -p[10] * X[:, 2] - p[11] * X[:, 2] - p[12] * X[:, 2]
    j[:, 4] = X[:, 0]
    j[:, 5] = X[:, 1]
    j[:, 6] = X[:, 2]
    j[:, 7] = -(p[10] + p[11] + p[12]) * 0.5 * X[:, 4]
    j[:, 8] = -(p[10] + p[11] + p[12]) * 0.5 * X[:, 5]
    j[:, 9] = -(p[10] + p[11] + p[12]) * 0.5 * X[:, 6]
    j[:, 10] = (X[:, 3] - (p[1] * X[:, 0] + p[2] * X[:, 1] + p[3] * X[:, 2]) -
                0.5 * (p[7] * X[:, 4] + p[8] * X[:, 5] + p[9] * X[:, 6]))
    j[:, 11] = (X[:, 3] - (p[1] * X[:, 0] + p[2] * X[:, 1] + p[3] * X[:, 2]) -
                0.5 * (p[7] * X[:, 4] + p[8] * X[:, 5] + p[9] * X[:, 6]))
    j[:, 12] = (X[:, 3] - (p[1] * X[:, 0] + p[2] * X[:, 1] + p[3] * X[:, 2]) -
                0.5 * (p[7] * X[:, 4] + p[8] * X[:, 5] + p[9] * X[:, 6]))
    return j

p_init = np.repeat(1, 13)
fit_ids_protein = optim.least_squares(fun_fit_ids, p_init, jac=jacobian_ids, args=(X_ids, y_ids, 'protein'))
fit_ids_fat = optim.least_squares(fun_fit_ids, p_init, jac=jacobian_ids, args=(X_ids, y_ids, 'fat'))
fit_ids_carb = optim.least_squares(fun_fit_ids, p_init, jac=jacobian_ids, args=(X_ids, y_ids, 'carb'))

print('Nonlinear IDS dependent: protein consumption; price effects:', np.round(fit_ids_protein.x[4:7], 4))
print()
print('Nonlinear IDS dependent: fat consumption; price effects:', np.round(fit_ids_fat.x[4:7], 4))
print()
print('Nonlinear IDS dependent: carb consumption; price effects:', np.round(fit_ids_carb.x[4:7], 4))

Nonlinear IDS dependent: protein consumption; price effects: [-903.5613   87.4024  491.8787]

Nonlinear IDS dependent: fat consumption; price effects: [ 964.4698 -557.3486 -849.9544]

Nonlinear IDS dependent: carb consumption; price effects: [-2465.7473    75.6284  -855.3498]


Price effect vector elements remain the same: protien in the first position, fat second, and carbohydrates third. Values are interpreted as gram units. We still see negative own-price effeccts but there is a difference in cross-price relationships depenendent on the macronutrient consumption variable. Fats and carbohydrates both show substitutable relationships with protein. When modeling fat consumption, protein is shown to be substitutable while carbohydrates is complimentary. Modeling carbohydrate consumption shows protein is complimentary while fat is substitutable.